File: Others/medium/SpiralTraverse.py
def spiralTraverse(array):

    num_elements = len(array) * len(array[0])

    n = len(array)
    m = len(array[0])
    it = 0
    result = []
    while num_elements > 0:
        # Up side
        for j in range(it, m - it):
            result.append(array[it][j])
            num_elements -= 1
            if num_elements == 0:
                break
        if num_elements == 0:
            continue

        # Right side
        for i in range(it + 1, n - it):
            result.append(array[i][m - 1 - it])
            num_elements -= 1
            if num_elements == 0:
                break
        if num_elements == 0:
            continue

        # Bottom side
        for j in reversed(range(it, m - 1 - it)):
            result.append(array[n - it - 1][j])
            num_elements -= 1
            if num_elements == 0:
                break
        if num_elements == 0:
            continue

        # Left side
        for i in reversed(range(it + 1, n - it - 1)):
            result.append(array[i][it])
            num_elements -= 1
            if num_elements == 0:
                break
        it += 1

    return result

File: Golang/medium/spiral-traverse.go
package main

func SpiralTraverse(array [][]int) []int {
	numElements := len(array) * len(array[0])

	n := len(array)
	m := len(array[0])
	it := 0
	result := []int{}

	for numElements > 0 {
		for j := it; j < m-it; j++ {
			result = append(result, array[it][j])
			numElements -= 1
			if numElements == 0 {
				break
			}
		}

		if numElements == 0 {
			continue
		}

		for i := it + 1; i < n-it; i++ {
			result = append(result, array[i][m-1-it])
			numElements -= 1
			if numElements == 0 {
				break
			}
		}

		if numElements == 0 {
			continue
		}

		for j := m - it - 2; j >= it; j-- {
			result = append(result, array[n-it-1][j])
			numElements -= 1
			if numElements == 0 {
				break
			}
		}

		if numElements == 0 {
			continue
		}

		for i := n - it - 2; i >= it+1; it-- {
			result = append(result, array[i][it])
			numElements -= 1
			if numElements == 0 {
				break
			}
		}

		it += 1
	}
	return result
}
